File: leetcode/Arrays/leetcode-0189/189.go
package arrays

func rotate(nums []int, k int) {
	if nums == nil || len(nums) == 0 {
		return
	}

	if k > len(nums) {
		k -= len(nums)
	}

	help(nums, 0, len(nums)-1)
	help(nums, 0, k-1)
	help(nums, k, len(nums)-1)
}

func help(nums []int, start, end int) {
	for start < end {
		nums[start], nums[end] = nums[end], nums[start]
		start++
		end--
	}
}

File: leetcode/Arrays/leetcode-0189/189.java
class Solution {
    public void rotate(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return;
        }
        help(nums, 0, nums.length - 1);
        help(nums, 0, k - 1);
        help(nums, k, nums.lenght - 1);
    }

    private void help(int[] nums, int start, int end) {
        while (start < end) {
            int tmp = nums[start];
            nums[start] = nums[end];
            nums[end] = tmp;
        }
    }
}